distributed/core.py

```diff
@@ -118,93 +128,113 @@ def __init__(self, handlers, max_buffer_size=MAX_BUFFER_SIZE,
                 self._tick_pc = PeriodicCallback(self._measure_tick, 20, io_loop=self.loop)
                 self.loop.add_callback(self._tick_pc.start)
 
-
         self.__stopped = False
 
-        super(Server, self).__init__(max_buffer_size=max_buffer_size, **kwargs)
-
     def stop(self):
         if not self.__stopped:
             self.__stopped = True
-            super(Server, self).stop()
+            if self.listener is not None:
+                self.listener.stop()
 
     def _measure_tick(self):
         now = time()
         self.digests['tick-duration'].add(now - self._last_tick)
         self._last_tick = now
 
     @property
+    def address(self):
+        """
+        The address this Server can be contacted on.
+        """
+        if not self._address:
+            if self.listener is None:
+                raise ValueError("cannot get address of non-running Server")
+            self._address = self.listener.contact_address
+        return self._address
+
+    @property
+    def listen_address(self):
+        """
+        The address this Server is listening on.  This may be a wildcard
+        address such as `tcp://0.0.0.0:1234`.
+        """
+        if not self._listen_address:
+            if self.listener is None:
+                raise ValueError("cannot get listen address of non-running Server")
+            self._listen_address = self.listener.listen_address
+        return self._listen_address
+
+    @property
     def port(self):
+        """
+        The port number this Server is listening on.
+        """
         if not self._port:
-            try:
-                self._port = first(self._sockets.values()).getsockname()[1]
-            except StopIteration:
-                raise OSError("Server has no port.  Please call .listen first")
+            _, loc = parse_address(self.address)
+            _, self._port = parse_host_port(loc)
```

Comment:
No, not always. This part isn't properly abstracted. One would have to review if the `self.port` attribute is still used, and whether/how to eliminate it...